### ISSUE TITLE
[Intern Progress] Antony Nyagah - Ersilia Outreachy Contributions

### ISSUE THREAD

**Issue by tony-nyagah** [ersilia-os/outreachy-contributions#13]
# 🌟 Intern Progress Tracker  
Welcome! Use this issue to track your progress as you work through the AI/ML modeling exercise. Follow each step, check the boxes when completed, and update your issue as you go.

# ▶️ **Entry tasks**
- [x] Motivation letter posted in this issue
- [x] Installed the Ersilia Model Hub and successfully run models!
- [x] Recorded system settigns in this issue
- [x] Explained completed tests of Ersilia in the issue

---

##  **1. Download a Dataset of Interest**
- [x] Created a fork of this repository to work on
- [x] Selected a dataset from Therapeutics Data Commons (TDC)
- [x] Verified that the dataset is a classification problem*
- [x] Read and understood the dataset’s background and endpoint  
- [x] Assessed computational feasibility (size, features, etc.)  
- [x] Downloaded the dataset using Python  
- [x] Saved data in the `/data` folder  
- [x] Started writing documentation in the README (Installation, dataset details, how to run code)  

---

##  **2. Featurise the Data**
- [x] Browsed Ersilia Model Hub for "Representation" models  
- [x] Selected an appropriate featuriser and justified my choice  
- [x] Implemented featurisation in a notebook or script
- [x] Ensured the process is reproducible 
- [x] Updated the README with featurisation instructions  

---

##  **3. Build an ML Model**
- [x] Selected an ML framework
- [x] Trained the model on featurised data  
- [x] Evaluated the model (accuracy, precision, recall, etc.)  
- [x] Created graphs using Matplotlib for better visualization  
- [x] Ensured all steps are reproducible  
- [x] Documented the model-building process in README

---

##  **4. Prepare Code for Review**
- [x] Double-checked that documentation is clear and complete  
- [x] Evaluated if my model performance is satisfactory 
- [x] Identified possible improvements  
- [x] Updated this issue to request a review 

---

##  **5. Stretch Tasks (Optional)**
- [ ] Tried a different **featuriser**  
- [ ] Tested other **ML architectures**  
- [ ] Applied the model to **public datasets (e.g., ChEMBL)**  
- [x] Any other improvements? (Describe below)  

---

## 📌 **Requesting Review**
Once you’ve completed **Steps 1-4**, request a review by mentioning a mentor:  
👤 *@GemmaTuron* Please review my work!  

---

### 📝 **Asking questions**
Use this issue to update your progress, explain where you are stuck or ask for review!

Thank you for contributing! 🎉

**Comment by tony-nyagah** [comment]
## Motivation Letter

Hello,

I am Antony Nyagah, a software developer and data enthusiast with an interest in healthcare related projects. I have gained most of my skills from maintaining and working on projects that help labs perform external quality assessment. Due to the nature of how I gained my skills and my interest in using technology to make healthcare more equitable, Ersilia's goals have caught my eye. I'd love to help in furthering the democratize of biomedical research with Ersilia.

My skills are mostly in Python and JavaScript with a focus on backend development and data. Apart from writing and maintaining code I can deploy said code using Docker and CICD pipelines on either GitLab or GitHub with minimal issues. I also have a data science certification from ALX Africa and a data analyst certification from Datacamp. This is the perfect opportunity to put what I learned into practice. Having run the gamut of software development, I feel I have the necessary skills to aid in furthering Ersilia's goals while also learning of how to further improve healthcare in my country and eventually worldwide.

My main goals are to learn how to implement models that produce accurate results, learn from the community and to implement what I have learned within the health sector in my country. The implementation part could be either through educating others on what Ersilia is doing and how it can help labs or actually getting down to it and working on some models with help from people far more knowledgeable than I am. 

I look forward to getting started on the contributions.

Best regards,
Antony Nyagah

**Comment by tony-nyagah** [comment]
## Install the Ersilia Model Hub and successfully run models!
I have successfully installed the Ersilia Model Hub by following the instructions at https://ersilia.gitbook.io/ersilia-book/ersilia-model-hub/installation.

The models I ran are `eos3b5e`, `eos2r5a`, and `eos9ei3`. These will be run on the Ibuprofen SMILES string as input so the command should be:
1. `ersilia fetch EOS-ID` - fetch the model we're going to use
2. `ersilia serve EOS-ID` - serve the model
3. `ersilia run -i "CC(C)CC1=CC=C(C=C1)C(C)C(=O)O"` - send the ibuprofen smiles string as input

The results for each model are as follow:
* `eos3b5e`: molecular-weight
![eos3b5e results for Ibuprofen](https://github.com/user-attachments/assets/6055dd1c-d96d-4b71-acf3-cd5b4e90e6cd)

* `eos2r5a`: retrosynthetic-accessibility
![eos2r5a results for Ibuprofen](https://github.com/user-attachments/assets/a26b34d6-7c7d-48d0-97ec-ffe28b94f14d)

* `eos9ei3`: sa-score
![eos9ei3 results for Ibuprofen](https://github.com/user-attachments/assets/733847d2-afae-4a66-9739-5063ca490e33)

**Comment by tony-nyagah** [comment]
## System settings
My system settings are as follows:

![System settings](https://github.com/user-attachments/assets/5d73a611-2740-4933-95dc-7f373ab2c9af)
-  **OS**: *Fedora*
- **Python**: *3.12.9*
- **Docker**: *28.0.1*
- **Docker** *Compose: 2.33.1*

**Comment by tony-nyagah** [comment]
## Explained completed tests of Ersilia
The models I ran can receive SMILES which are defined as:

> Simplified Molecular Input Line Entry System (SMILES) which are used to translate a chemical's three-dimensional structure into a string of symbols that is easily understood by computer software.
> --- https://www.epa.gov/sites/default/files/2015-05/documents/appendf.pdf

After installing Ersilia, access to ready-to-use machine learning models for drug discovery and global health is gained.

The models to run for the task were:
1. `eos3b5e` - Molecular weight
    - The model is simply an implementation of the function Descriptors.MolWt of the chemoinformatics package RDKIT. It takes as input a small molecule (SMILES) and calculates its molecular weight in g/mol.
    - ![Molecular weight info](https://github.com/user-attachments/assets/9c9d60c5-bde9-4738-b196-34e78a9a2f61)

    - The molecular weight of Ibuprofen ended up being `206.28499999999997`.

2. `eos2r5a` - Retrosynthetic accessibility score
    - Retrosynthetic accessibility score based on the computer aided synthesis planning tool AiZynthfinder. The authors have selected a ChEMBL subset of 200.000 molecules, and checked whether AiZinthFinder could identify a synthetic route or not. This data has been trained to create a classifier that computes 4500 times faster than the underlying AiZynthFinder. Molecules outside the applicability domain, such as the GBD database, need to be fine tuned to their use case.
    - ![Retrosynthetic accessibility score](https://github.com/user-attachments/assets/82e0a473-afd2-4a0a-94f9-906bd547d343)
    - The retrosynthetic accessibility score of Ibuprofen ended up being: `0.9899624`.

3. `eos9ei3` - Synthetic accessibility score
    - Estimation of synthetic accessibility score (SAScore) of drug-like molecules based on molecular complexity and fragment contributions. The fragment contributions are based on a 1M sample from PubChem and the molecular complexity is based on the presence/absence of non-standard structural features. It has been validated comparing the SAScore and the estimates of medicinal chemist experts for 40 molecules (r2 = 0.89). The SAScore has been contributed to the RDKit Package.
    - ![Synthetic accessibility score](https://github.com/user-attachments/assets/66a9013a-5a49-4dc4-afbe-e2a7f42ccbaa)
    - The synthetic accessibility score of Ibuprofen ended up being: `2.191755`.

To view what models I have downloaded I ran `ersilia catalog`:
![Ersilia catalog command results](https://github.com/user-attachments/assets/1b0c3d5b-0f77-469b-b83d-68ca47f08189)

The `ersilia fetch` command uses Docker to get the required image and create containers:
![Ersilia model Docker containers](https://github.com/user-attachments/assets/e04de8b9-a3ec-47ee-8cfa-0bc040c0d443)

The `fetch` command can be used to pull from various sources like s3 and DockerHub but you can also just use Docker itself to pull models you need:
![Pulling models with Docker](https://github.com/user-attachments/assets/e2a9abeb-28b8-4bb1-9b04-c1ef5e218664)

**Comment by tony-nyagah** [comment]
@GemmaTuron Please review my work. Thanks.

**Comment by GemmaTuron** [comment]
Hello @tony-nyagah 

Thanks for this good start to the contribution period, great to have you on board and to hear you've already worked with Docker and CI/CD pipelines
Please go ahead and register your first contribution for Ersilia in the Outreachy page linking this issue, where we will continue working during the rest of the internship. If you can, join in the welcome call today at 17:00 CET for more info!

**Comment by tony-nyagah** [comment]
> Hello [@tony-nyagah](https://github.com/tony-nyagah)
> 
> Thanks for this good start to the contribution period, great to have you on board and to hear you've already worked with Docker and CI/CD pipelines Please go ahead and register your first contribution for Ersilia in the Outreachy page linking this issue, where we will continue working during the rest of the internship. If you can, join in the welcome call today at 17:00 CET for more info!

Thanks for the feedback. Very much appreciated.

**Comment by GemmaTuron** [comment]
Hi @tony-nyagah 

If you plan on making an application to Ersilia please go ahead and get started with w2 and 3 tasks so you have time to complete them

**Comment by tony-nyagah** [comment]
> Hi [@tony-nyagah](https://github.com/tony-nyagah)
> 
> If you plan on making an application to Ersilia please go ahead and get started with w2 and 3 tasks so you have time to complete them

Hi @GemmaTuron. I have what I am working on on my machine all I need is to push the code.    
I will notify you when I push the changes between today and tomorrow. Is that okay with you?

**Comment by tony-nyagah** [comment]
Hello @GemmaTuron 

I have update my fork and submitted a PR with my Outreachy submission: https://github.com/ersilia-os/outreachy-contributions/pull/53

I'd greatly appreciate your feedback.

**Comment by GemmaTuron** [comment]
Hi @tony-nyagah 

In order for me to revise the work please add a summary on this issue on the findings, results and so of your project. If you are at a loss check your peers issues to see the kind of summary they provided.
Thanks

**Comment by tony-nyagah** [comment]
Here's a summary of what I have done so far:
1. Created an easy way to get data from [tdcommons](https://tdcommons.ai/) using the following script setup:
    - `python scripts/fetch_dataset.py --dataset_name <dataset_name> --dataset_group <dataset_group>`
    - This enables fetching any dataset as long as you have the dataset name and group.
2. Created a script for featurizing the data you've downloaded using a passed in Ersilia model:
    - `python scripts/featurization.py --dataset_name <dataset_name> --model_id <ersilia_model_id> --feature_name <feature_name>`
    - This script adds a column with the passed in feature name to all the CSVs related to the dataset you want to featurize.
    - If the Ersilia model returns multiple values, we can exclude the `--feature-name` so that the additional column names are generated straight from the model's results e.g. `python scripts/featurization.py --dataset_name AMES --model_id eos9gg2
`
3. Added features to the AMES dataset I'm using by taking advantage of various Ersilia models including: `sa-score`, `molecular-weight` and `chemical-space-projections-drugbank`.
4. Cleaned data. Some models like the one for molecular weight returned lists instead of actual numbers which affected data modelling.
5. Trained machine learning models and chose the best one. I used Random Forest classifier, XGBoost Classifier and handled class imbalances with SMOTE.
6. Generated various visualizations to better understand and display the effectiveness of the process of creating the model.

My final results were as follows for the models:
    - Random Forest: 75.76%
    - Tuned Random Forest: 75.76%
    - Random Forest with SMOTE: 73.42%
    - XGBoost: 70.80%
    - XGBoost with SMOTE: 69.01%

The standard Random Forest model was selected as the best model based on validation accuracy. Its performance on the test set was:
    - Accuracy: 74.21%
    - Precision: 75.15%
    - Recall: 79.02%
    - F1 Score: 77.04%

I finally worked on a simple web app to  display model results. The app hasn't been deployed but can be run manually using after installing dependencies in `environment.yml`:
```bash
cd web_app
uvicorn main:app --reload
```

![Image](https://github.com/user-attachments/assets/942082e9-0326-4e56-b2ef-7e5959f6aa37)

**Comment by tony-nyagah** [comment]
@GemmaTuron I have updated the issue with what I have managed to do. Sorry if it's a bit long.

It has been a good learning opportunity. I've learned that I should work on better understanding how to create features. Will be using Ersilia and tdcommons to this end.

**Comment by GemmaTuron** [comment]
Thanks @tony-nyagah for the work!

I am assuming this: _My final results were as follows for the models_ is referring to an AUROC? please make sure all the project steps are well explained and the results as well to submit the final application. 
Once you are happy with it go ahead and start working on the questions posed in the Outreachy final application. If you have attempted any model refactoring please link it here as well

**Comment by tony-nyagah** [comment]
> Thanks @tony-nyagah for the work!
> 
> I am assuming this: _My final results were as follows for the models_ is referring to an AUROC? please make sure all the project steps are well explained and the results as well to submit the final application. 
> Once you are happy with it go ahead and start working on the questions posed in the Outreachy final application. If you have attempted any model refactoring please link it here as well

Yes, the results refer to an AUROC. I'll make my project steps clearer then submit my final application.

Thanks for the feedback.

**Comment by miquelduranfrigola** [comment]
Auto-closing this Outreachy intern-progress tracker as completed — the contribution round has ended. Reopen if follow-up is needed.
